### ISSUE TITLE
!fatal 70 - Illegal property of an endo relation 

### ISSUE THREAD

**Issue by RieksJ** [AmpersandTarski/Ampersand#383]
Consider the following script
~~~
CONTEXT Bug

CLASSIFY Term ISA Variable
varIsUsedBy :: Variable * Term [UNI,IRF,ASY]

ENDCONTEXT
~~~

This produces the following fatal error:
~~~
ampersand.exe: !fatal, called at src\Database\Design\Ampersand\ADL1\Rule.hs:55:22 in main:Database.Design.Ampersand.ADL1.Rule
             Ampersand v3.5.0[branch_#228:e920f70]
             error nr: 70
  Illegal property of an endo relation "varIsUsedBy"
~~~

This is a bug because we should not see fatals.

I wonder why we could not specify this - at least intuitively, it has meaning...

**Comment by WolframKahl** [comment]
On Fri, Apr 15, 2016 at 02:36:26PM -0700, Rieks wrote:
> Consider the following script
> ~~~
> CONTEXT Bug
> 
> CLASSIFY Term ISA Variable
> varIsUsedBy :: Variable * Term [UNI,IRF,ASY]
> 
> ENDCONTEXT
> ~~~


TODO: Implement semantic processing and a mathematical terminology knowledge base, so that Ampersand can already throw a fatal error at

> CLASSIFY Term ISA Variable

as conceptually dangerous...

:wink:




More seriously:

> This produces the following fatal error:
> ~~~
> ampersand.exe: !fatal, called at src\Database\Design\Ampersand\ADL1\Rule.hs:55:22 in main:Database.Design.Ampersand.ADL1.Rule
>              Ampersand v3.5.0[branch_#228:e920f70]
>              error nr: 70
>   Illegal property of an endo relation "varIsUsedBy"
> ~~~
> 
> This is a bug because we should not see fatals.

(Yes.)

> I wonder why we could not specify this - at least intuitively, it has meaning...

With the ```IRF``` in

> varIsUsedBy :: Variable * Term [UNI,IRF,ASY]

, you are probably thinking (how do you write empty relations?):

```
RULE: I[Term] /\ varIsUsedBy |- EMPTY
```

, which looks like irreflexivity, except that it is not about homogeneous relations.

Using
```
TinjV :: Term * Variable [UNI,TOT,INJ]
```
for the injection behind
```
CLASSIFY Term ISA Variable
```
, the rule you intended then becomes:
```
RULE: TinjV~ ; I[Term] /\ varIsUsedBy |- EMPTY[Variable * Term]
```
or simpler:```
RULE: TinjV~ /\ varIsUsedBy |- EMPTY[Variable * Term]
```
It translates into English as:
``each term, considered as a variable, is not used by itself''.
Which brings me right back to ``conceptually dangerous''...


Cheers,

Wolfram

**Comment by WolframKahl** [comment]
(Redoing this since github didn't display half of my e-mail reply ─ sorry for the noise! I also added a bit more at the bottom.)

With the ```IRF``` in

```
varIsUsedBy :: Variable * Term [UNI,IRF,ASY]
```
, you are probably thinking (how do you write empty relations?):

```
RULE: I[Term] /\ varIsUsedBy |- EMPTY
```

, which looks very similar to irreflexivity, but it is not about homogeneous relations.

Using

```
TinjV :: Term * Variable [UNI,TOT,INJ]
```

for the injection behind

```
CLASSIFY Term ISA Variable
```

, the rule you intended then becomes:

```
RULE: TinjV~ ; I[Term] /\ varIsUsedBy |- EMPTY[Variable * Term]
```

or simpler:

```
RULE: TinjV~ /\ varIsUsedBy |- EMPTY[Variable * Term]
```

It translates into English as:
“each term, considered as a variable, is not used by itself (considered as a term)”.
Which brings me right back to “conceptually dangerous”...

If you want to think about
```
RULE: I[Term] /\ varIsUsedBy |- EMPTY
```
as being about homogeneous relations, then you have to consider it on ```Variable * Variable```, since ```Variable``` is the larger concept (in your warped ontology), so now it is:
```
RULE: TinjV~ ; I[Term] ; TinjV /\ varIsUsedBy ; TinjV |- EMPTY[Variable * Variable]
```
This is now quite clearly NOT irreflexivity.

**Comment by stefjoosten** [comment]
This is an omission. There should be a static error message to signal this at compile time.

Thank you, Wolfram for pointing out why we want to restrict the shorthand notation `IRF` to endorelations.

By the way, `-V` is how we write the empty relation. It fits nicely in the type system: if the type system cannot derive its type, you must add it just as in other expressions. And yes, this means we have a lot of different empty sets. Just as many as there are full sets `V`.

**Comment by hanjoosten** [comment]
Fixes in commit 563239d9fa9887a17e5daf70dd35d2e00914c0ce
